// @rampant/ai/test-intents.js
import { white, green, grey } from 'chalk'

// eslint-disable-next-line import/no-mutable-exports
export let testCount = 0
const success = msg => {
    testCount += 1
    return `    ${white(`#${testCount}`.padEnd(3, ' '))} ${green('✔')} ${grey(
        msg
    )}`
}

export const project = {
    intents: [
        {
            name: 'logTest',
            effects: [{ type: 'log', message: success('Log effects work') }],
        },
        {
            name: 'vote',
            defaults: {
                doc: 'votes',
                data: { dusty: 1 },
            },
            effects: [
                { type: 'db', doc: '{$.params.doc}', data: '{$.params.data}' },
                { type: 'log', message: success('Database effects work') },
            ],
        },
        {
            name: 'voteResults',
            sources: [{ type: 'db', name: 'votes', intent: 'voteResults' }],
            defaults: {
                doc: 'vote-results',
                data: { dusty: 1 },
            },
            effects: [
                { type: 'db', doc: '{$.params.doc}', data: '{$.params.data}' },
                { type: 'log', message: success('Database sources work') },
            ],
        },
        {
            name: 'staticPipeline',
            effects: [{ type: 'log', message: success('Static pipes work') }],
        },
        {
            name: 'basicTemplateTest',
            defaults: {
                greeting: 'Hello',
                subject: 'World',
            },
            pipeline: [
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        'Basic templating works ({{$.params.greeting}} {{$.params.subject}}!)'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'advancedTemplateTest',
            defaults: {
                votes: {
                    raymond: 'dusty',
                    jessica: 'retail',
                    stephanie: 'dusty',
                },
            },
            pipeline: [
                {
                    type: 'static',
                    key: 'results',
                    value: `{% const tally = votes =>
                    JSON.stringify(
                        Object.entries(votes).reduce(
                            (all, [, vote]) => ({
                                ...all,
                                [vote]: vote in all ? all[vote] + 1 : 1,
                            }),
                            {}
                        )
                    );
                %}{{ tally($.params.votes) }}`,
                },
                {
                    type: 'static',
                    key: 'message',
                    value: success('Advanced templating works ({{$.results}})'),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'parseIntTest',
            defaults: {
                first: 2,
                second: 3,
            },
            pipeline: [
                {
                    type: 'static',
                    key: 'total',
                    parse: 'number',
                    value: `{{ $.params.first + $.params.second }}`,
                },
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        'Parsing as an integer works ({{ $.total }} is of type {{ typeof $.total }})'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'parseObjectTest',
            defaults: {
                stringified: '{"dusty":2,"retail":1}',
            },
            pipeline: [
                {
                    type: 'static',
                    key: 'object',
                    parse: 'object',
                    value: `{{ $.params.stringified }}`,
                },
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        'Parsing as an object works ({{ $.params.stringified }} is of type {{ typeof $.object }})'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'testMixerMessage',
            defaults: { message: success('Mixer Message Effect works') },
            effects: [
                {
                    type: 'message',
                    text: 'Hello',
                    userId: 'muonglow',
                    channelId: '41469686',
                },
                {
                    type: 'message',
                    text: '!dep',
                    userId: 'muonglow',
                    channelId: '41469686',
                },
                { type: 'log', message: '{$.params.message}' },
            ],
        },
        {
            name: 'mixerChatTest',
            sources: [{ name: 'botChat' }],
            pipeline: [
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        'Mixer Chat works ({{ $.params.rawMessage }})'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'fromTest',
            sources: [
                {
                    type: 'from',
                    name: 'from',
                    intent: 'fromTest',
                    items: [{ item: 'First' }, { item: 'Second' }],
                },
            ],
            pipeline: [
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        '{{ $.params.item }} message for From source works'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'filterTest',
            sources: [
                {
                    type: 'from',
                    name: 'filter',
                    intent: 'filterTest',
                    items: [
                        { message: 'Not command' },
                        { message: '!command' },
                    ],
                },
            ],
            pipeline: [
                {
                    type: 'filter-regex',
                    regex: '^[.!,\\s]+(.*)',
                    field: '{$.params.message}',
                },
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        '{{ $.params.message }} was filtered properly'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'sourceDependencyTest',
            aliases: ['depTest'],
            sources: [{ name: 'commands' }],
            pipeline: [
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        'Command dependency source works ({{ $.params.rawMessage }})'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'greeting',
            samples: ['Hello', 'Hi', 'Hey'],
            sources: [{ name: 'nlp' }],
            pipeline: [
                {
                    type: 'static',
                    key: 'message',
                    value: success(
                        'NLP works ("{{ $.params.rawMessage }}" is of intent: {{ $.intent }})'
                    ),
                },
            ],
            effects: [{ type: 'log', message: '{$.message}' }],
        },
        {
            name: 'twitterRetweetTest',
            defaults: {
                userId: 'raybenefield',
                tweetId: '1166372241149329408',
            },
            effects: [
                /*
                {
                    type: 'retweet',
                    userId: '{$.params.userId}',
                    tweetId: '{$.params.tweetId}',
                },
                */
                { type: 'log', message: success('Retweet works') },
            ],
        },
        {
            name: 'filterMentionsTest',
            sources: [{ name: 'botChat' }],
            //effects: [{ type: 'log', message: '{$}' }],
        },
    ],
}

export const cleanupDocs = ['votes', 'vote-results']
